### ISSUE TITLE
cli/server: "--split-mode row" breaking with multiple AMD GPU's - ROCm; incredibly slow without it

### ISSUE THREAD

**Issue by mpapili** [ggerganov/llama.cpp#10682]
### Name and Version
```
# ./llama-server --version
ggml_cuda_init: found 2 ROCm devices:
  Device 0: AMD Radeon RX 6800, compute capability 10.3, VMM: no
  Device 1: AMD Radeon RX 6700 XT, compute capability 10.3, VMM: no
version: 4240 (64ed2091)
built with cc (Ubuntu 13.2.0-23ubuntu4) 13.2.0 for x86_64-linux-gnu
```

### Operating systems

Linux (Ubuntu 24.04 LTS)

### Which llama.cpp modules do you know to be affected?

llama-cli, llama-server

### Problem description & steps to reproduce

When built with hipBLAS to use with ROCm there is a significant slowdown when not using `--split-mode`

The slowdown is as such:
```
6800 alone:  llama_perf_context_print 54.47 tokens per second
6700xt alone: llama_perf_context_print 45.47 tokens per second
combined without row-split: 25.54 tokens per second (significant slowdown)
combined with row-split 38.99 tokens per second (but nonsense responses)
```

Running with split-mode gets me pretty close to the inference speed of the slower GPU (the Rx 6700xt, with ~300gb/s memory) however running with:
```
 HSA_OVERRIDE_GFX_VERSION=10.3.0 ./llama-cli \
        -m ${LLM_DIR}/${LLAMA3_8B}  \
        --ctx-size 2000 \
	-p "who is the president of Poland?" \
	-n 50 \
	-ngl 200 \
	--split-mode row 
```
yields a response that is usually nonsense, most-often the letter `G` repeated endlessly for the length of the response. Combinations of `batch_size` and `ubatch_size` do not appear to work either.

### Relevant log output

```shell
llama_new_context_with_model: n_ctx_per_seq (2016) < n_ctx_train (131072) -- the full capacity of the model will not be utilized
llama_kv_cache_init:      ROCm0 KV buffer size =   149.62 MiB
llama_kv_cache_init:      ROCm1 KV buffer size =   102.38 MiB
llama_new_context_with_model: KV self size  =  252.00 MiB, K (f16):  126.00 MiB, V (f16):  126.00 MiB
llama_new_context_with_model:  ROCm_Host  output buffer size =     0.49 MiB
llama_new_context_with_model:      ROCm0 compute buffer size =   323.88 MiB
llama_new_context_with_model:      ROCm1 compute buffer size =   517.00 MiB
llama_new_context_with_model:  ROCm_Host compute buffer size =    23.89 MiB
llama_new_context_with_model: graph nodes  = 1030
llama_new_context_with_model: graph splits = 3
common_init_from_params: warming up the model with an empty run - please wait ... (--no-warmup to disable)
main: llama threadpool init, n_threads = 16

system_info: n_threads = 16 (n_threads_batch = 16) / 32 | ROCm : PEER_MAX_BATCH_SIZE = 128 | CPU : SSE3 = 1 | SSSE3 = 1 | AVX = 1 | AVX2 = 1 | F16C = 1 | FMA = 1 | LLAMAFILE = 1 | AARCH64_REPACK = 1 | 

sampler seed: 2080410442
sampler params: 
	repeat_last_n = 64, repeat_penalty = 1.000, frequency_penalty = 0.000, presence_penalty = 0.000
	dry_multiplier = 0.000, dry_base = 1.750, dry_allowed_length = 2, dry_penalty_last_n = -1
	top_k = 40, top_p = 0.950, min_p = 0.050, xtc_probability = 0.000, xtc_threshold = 0.100, typical_p = 1.000, temp = 0.800
	mirostat = 0, mirostat_lr = 0.100, mirostat_ent = 5.000
sampler chain: logits -> logit-bias -> penalties -> dry -> top-k -> typical -> top-p -> min-p -> xtc -> temp-ext -> dist 
generate: n_ctx = 2016, n_batch = 1024, n_predict = 200, n_keep = 1

who is the president of poland?GGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG
```
```

**Comment by JohannesGaessler** [comment]
Which exact model are you using?

**Comment by mpapili** [comment]
> Which exact model are you using?

Tested this with a few models:
- Meta-Llama-3.1-8b_Q6_K_L.gguf (the test that produced the benchmarks
- Mistral-Small-22b_Q6_K_M.gguf
- Codestral-22b_Q6_K_L.gguf
- Codestral-22b_Q4_M.gguf

The numbers in my original post were for Meta-Llama-3.1-8b_Q6_K_L.gguf but I can recreate the behavior on all other models I try.

**Comment by JohannesGaessler** [comment]
I am not seeing any issues on my setups with multiple NVIDIA GPUs (I don't have one with multiple AMD GPUs). Does any of the following lead to correct results:

* Setting `-ub 1` for the server.
* Compiling with `GGML_CUDA_FORCE_CUBLAS=ON`.
* Compiling with `GGML_CUDA_PEER_MAX_BATCH_SIZE=0`.
* Using an FP16 model.

**Comment by mpapili** [comment]
Thank you for the suggestion, I will test the above when back at my desktop and get back to you on the results. 

One more data point to add from testing this morning: running with Vulkan yields the same slowdown with multiple AMD GPUs, although `--split-mode row` does not fix the significant speed loss (~24t/s in the same test as my original post) it *does* return a normal-looking output.

**Comment by mpapili** [comment]
Okay a few more results to share:

Compiling with `-DGGML_CUDA_FORCE_CUBLAS=ON` AND `-DGGML_CUDA_PEER_MAX_BATCH_SIZE=0` options during cmake resulted in the same experience on all models and option combinations.

Running an FP16 model this way (Llama3.2-3B-F16.gguf):
```
f16 llama3-3b:
rx 6800 only:
llama_perf_context_print: prompt eval time =     140.93 ms /     9 tokens (   15.66 ms per token,    63.86 tokens per second)
llama_perf_context_print:        eval time =     664.51 ms /    38 runs   (   17.49 ms per token,    57.18 tokens per second)

rx 6700xt only:
llama_perf_context_print: prompt eval time =     170.32 ms /     9 tokens (   18.92 ms per token,    52.84 tokens per second)
llama_perf_context_print:        eval time =    2120.77 ms /    99 runs   (   21.42 ms per token,    46.68 tokens per second)

both (split mode layer):
llama_perf_context_print: prompt eval time =     118.46 ms /     9 tokens (   13.16 ms per token,    75.97 tokens per second)
llama_perf_context_print:        eval time =    1713.02 ms /    43 runs   (   39.84 ms per token,    25.10 tokens per second)
(NOTE: significant slowdown)

both (split-mode row):
llama_perf_context_print: prompt eval time =     205.54 ms /     9 tokens (   22.84 ms per token,    43.79 tokens per second)
llama_perf_context_print:        eval time =    2419.82 ms /    99 runs   (   24.44 ms per token,    40.91 tokens per second)
(NOTE: output is just a series of the character "G" repeating)
```

similar results with Vulkan in this case. It seems that using `--split-mode row` gets us at the expected speed (slightly slower than the slower of the two GPU's) - however the output is nonsense, in this case:

```
who is the president of poland?GGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG
```

**Comment by JohannesGaessler** [comment]
My top priority in terms of debugging is determining why the results are wrong, performance numbers are right now not relevant because those are only valid if the results are also correct. To be clear, was there any combination of parameters for which `--split-mode row` resulted in intelligible model outputs regardless of performance?

**Comment by mpapili** [comment]
Got it - and no, I haven't yet found a way to use `--split-mode row` and get intelligible model outputs with ROCm (built enabling hipBLAS).

**Vulkan** does appear to work with `--split-mode row` on the same hardware setup however (though the performance benefits are not there)

**Comment by mpapili** [comment]
Hey so I got this figured out and wanted to share my solution here. It **is not** a bug/issue with llama-cpp, however this is something that I found multiple other users attempting to use multiple AMD GPU's were encountering, so I want to share my fix here:

**Fix:**
```
- enter BIOS
- disable CSM
    - perform a test-boot
- re-enter BIOS
- enable >4g Decoding
- switch pcie mode from 16x to 8x8 (may not be required for all motherboards, but was needed for my x470 Fatal1ty)
- enable BAR resizing
```

This can be closed. User configuration error - but hopefully one that can help others.
